### DISCUSSION TITLE
how to prevent stretch layout from jumping on content changes

### DISCUSSION THREAD

**Discussion by tjallingt** [slint-ui/slint#2890] [Q&A] [answered]
This is adapted from the widget galley demo:

```
import { CheckBox, GroupBox, SpinBox, ComboBox, VerticalBox, HorizontalBox } from "std-widgets.slint";

export component App inherits Window {
    preferred-height: 100px;
    preferred-width: 600px;

    VerticalBox {
        alignment: start;

        GroupBox {
            title: "CheckBox - SpinBox - ComboBox";

            HorizontalBox {
                alignment: stretch;

                checkbox := CheckBox {
                    horizontal-stretch: 1;

                    text: checkbox.checked ? "(checked)" : "(unchecked)";
                    checked: true;
                }


                SpinBox {
                    horizontal-stretch: 1;

                    value: 42;
                }

                ComboBox {
                    horizontal-stretch: 1;

                    model: ["Select Something", "From this", "Combobox"];
                }
            }           
        }
    }
}
```

If the checkbox is checked and unchecked the layout jumps because of the change in text.
I would like to prevent this jumping by having the `CheckBox` `SpinBox` and `ComboBox` take up 33% of the available width, and i expected the `horizontal-stretch` property to do that.
Given that the layout still jumps with the  `horizontal-stretch` it isn't clear to me how to prevent the layout shifting when the checkbox text changes.

**Comment by ogoffart** [comment, marked as answer]
You can use fixed with like this `width:33%;`

[example](https://slint.dev/releases/1.0.2/editor/?snippet=import+%7B+CheckBox%2C+GroupBox%2C+SpinBox%2C+ComboBox%2C+VerticalBox%2C+HorizontalBox+%7D+from+%22std-widgets.slint%22%3B%0A%0Aexport+component+App+inherits+Window+%7B%0A++++preferred-height%3A+100px%3B%0A++++preferred-width%3A+600px%3B%0A%0A++++VerticalBox+%7B%0A++++++++alignment%3A+start%3B%0A%0A++++++++GroupBox+%7B%0A++++++++++++title%3A+%22CheckBox+-+SpinBox+-+ComboBox%22%3B%0A%0A++++++++++++HorizontalBox+%7B%0A++++++++++++++++alignment%3A+stretch%3B%0A%0A++++++++++++++++checkbox+%3A%3D+CheckBox+%7B%0A++++++++++++++++++++horizontal-stretch%3A+1%3B%0A%0A++++++++++++++++++++text%3A+checkbox.checked+%3F+%22%28checked%29%22+%3A+%22%28unchecked%29%22%3B%0A++++++++++++++++++++checked%3A+true%3B%0A++++++++++++++++++++width%3A+33%25%3B%0A++++++++++++++++%7D%0A%0A%0A++++++++++++++++SpinBox+%7B%0A++++++++++++++++++++horizontal-stretch%3A+1%3B%0A++++++++++++++++++++value%3A+42%3B%0A++++++++++++++++++++width%3A+33%25%3B%0A++++++++++++++++%7D%0A%0A++++++++++++++++ComboBox+%7B%0A++++++++++++++++++++%2F%2Fwidth%3A+33%25%3B%0A++++++++++++++++++++horizontal-stretch%3A+1%3B%0A++++++++++++++++++++model%3A+%5B%22Select+Something%22%2C+%22From+this%22%2C+%22Combobox%22%5D%3B%0A++++++++++++++++%7D%0A++++++++++++%7D+++++++++++%0A++++++++%7D%0A++++%7D%0A%7D) 

(you can also specify manually the preferred-width and minimum-width.
If you set them all to the same value, then everything should take 1/3 of the available width.)

**Reply by tjallingt** [reply]
I'm okay with setting `width: 33%` but I would have expected `horizontal-stretch: 1;` to be enough for the components to automatically stretch to 33%. When adding a new component to the row it would be necessary to update all the explicitly specified widths.

**Reply by ogoffart** [reply]
the stretch decide how much they grow or shrink compared to their preferred size.
Since not all the element have the same preferred size, they are then having different sizes

**Reply by aurexav** [reply]
Not work with `space: 5px`, the last item will overflow.

`33% - space * (count - 1)` isn't a valid expression.

`parenet.width * 0.33 - space * (count - 1)` not work either(due to loop binding).
